cpp/src/arrow/flight/types.cc

```diff
@@ -463,6 +463,300 @@ arrow::Result<CancelFlightInfoRequest> CancelFlightInfoRequest::Deserialize(
+
+std::string GetSessionOptionsResult::ToString() const {
+  std::stringstream ss;
+  
+  ss << "<GetSessionOptionsResult session_options=" << session_options << '>';
+
+  return ss.str();
+}
+
+bool GetSessionOptionsResult::Equals(const GetSessionOptionsResult& other) const {
+  return CompareSessionOptionMaps(session_options, other.session_options);
+}
+
+arrow::Result<std::string>
+GetSessionOptionsResult::SerializeToString() const {
+  pb::GetSessionOptionsResult pb_result;
+  RETURN_NOT_OK(internal::ToProto(*this, &pb_result));
+
+  std::string out;
+  if (!pb_result.SerializeToString(&out)) {
+    return Status::IOError("Serialized GetSessionOptionsResult exceeded 2GiB limit");
+  }
+  return out;
+}
+
+arrow::Result<GetSessionOptionsResult>
+GetSessionOptionsResult::Deserialize(std::string_view serialized) {
+  pb::GetSessionOptionsResult pb_result;
+  if (serialized.size() > static_cast<size_t>(std::numeric_limits<int>::max())) {
+    return Status::Invalid(
+        "Serialized GetSessionOptionsResult size should not exceed 2 GiB");
+  }
+  google::protobuf::io::ArrayInputStream input(serialized.data(),
+                                               static_cast<int>(serialized.size()));
+  if (!pb_result.ParseFromZeroCopyStream(&input)) {
+    return Status::Invalid("Not a valid GetSessionOptionsResult");
+  }
+  GetSessionOptionsResult out;
+  RETURN_NOT_OK(internal::FromProto(pb_result, &out));
+  return out;
+}
+
+// CloseSessionRequest
+
+std::string CloseSessionRequest::ToString() const {
+  return "<CloseSessionRequest>";
+}
+
+bool CloseSessionRequest::Equals(const CloseSessionRequest& other) const {
+  return true;
+}
+
+arrow::Result<std::string>
+CloseSessionRequest::SerializeToString() const {
+  pb::CloseSessionRequest pb_request;
+  RETURN_NOT_OK(internal::ToProto(*this, &pb_request));
+
+  std::string out;
+  if (!pb_request.SerializeToString(&out)) {
+    return Status::IOError("Serialized CloseSessionRequest exceeded 2GiB limit");
+  }
+  return out;
+}
+
+arrow::Result<CloseSessionRequest>
+CloseSessionRequest::Deserialize(std::string_view serialized) {
+  pb::CloseSessionRequest pb_request;
+  if (serialized.size() > static_cast<size_t>(std::numeric_limits<int>::max())) {
+    return Status::Invalid(
+        "Serialized CloseSessionRequest size should not exceed 2 GiB");
+  }
+  google::protobuf::io::ArrayInputStream input(serialized.data(),
+                                               static_cast<int>(serialized.size()));
+  if (!pb_request.ParseFromZeroCopyStream(&input)) {
+    return Status::Invalid("Not a valid CloseSessionRequest");
+  }
+  CloseSessionRequest out;
+  RETURN_NOT_OK(internal::FromProto(pb_request, &out));
+  return out;
+}
+
+// CloseSessionResult
+
+std::string CloseSessionResult::ToString() const {
+  std::stringstream ss;
+
+  ss << "<CloseSessionResult result=" << status << '>';
+
+  return ss.str();
+}
+
+bool CloseSessionResult::Equals(const CloseSessionResult& other) const {
```

Comment:
There's alot of boiler plate. Are all these methods necessary? Can this be refactored to use templates?

Reply:
There should be a superclass for a bunch of these and templates for others; refactoring the existing code plus this PR's code is probably a separate task to get alignment on proper design there but agreed lots of things should be factored out, especially Deserialize/SerializeToString.